File: algorithms/cpp/583.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    int minDistance(string word1, string word2) {
        int m = word1.size(), n = word2.size();
        vector<vector<int>> dp(m+1, vector<int>(n+1));
        for (int i = 0; i <= m; i++) {
            for (int j = 0; j <= n; j++) {
                if (i == 0 || j == 0)
                    dp[i][j] = 0;
                else
                    dp[i][j] = (word1[i-1] == word2[j-1]) ? dp[i-1][j-1]+1
                             : max(dp[i-1][j], dp[i][j-1]);
            }
        }
        int commonLen = dp[m][n];
        return m - commonLen + n - commonLen;
    }
};

int main() {
    return 0;
}

File: algorithms/python/583.py
from bisect import bisect_left


class Solution:
    def buildArr(self, word1, word2):
        m = {}
        for idx, ch in enumerate(word2):
            if ch not in m:
                m[ch] = []
            m[ch].append(idx)
        arr = []
        for ch in word1:
            if ch in m:
                arr += m[ch][::-1]
        return arr

    def lis(self, arr):
        if not arr:
            return 0
        dp = [arr[0]]
        for num in arr:
            idx = bisect_left(dp, num)
            if idx == len(dp):
                dp.append(num)
            else:
                dp[idx] = num
        return len(dp)

    def minDistance(self, word1: str, word2: str) -> int:
        arr = self.buildArr(word1, word2)
        rls = self.lis(arr)
        return len(word1) + len(word2) - 2 * rls
